feat: redesign select dropdown

RutamBhagat opened the pull request:
## What does this PR do?
Update Select.tsx UI
/claim #4246

## Related issues
Fixes #4246

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Enhancement (non-breaking small changes to existing functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Explanation of the changes
Updated Select.tsx as requested.
Had to make changes to the props since pre-commit hooks were preventing from making a commit, please let me know if I need to make any other changes

RutamBhagat commented:
https://github.com/heyxyz/hey/assets/72187009/7fb81ce7-1577-48a1-9d12-305b6a2f5f72

Here is a demo

bigint commented:
Can you show demo with description as well, also trying all props like disabled, selected etc

also looks like spacing is very huge between two items!

RutamBhagat commented:
and for disabled options I have disabled the active state when hovered also disabled options cant be selected by default or when performing onClick on options

https://github.com/heyxyz/hey/assets/72187009/d06ed050-6851-4700-bac5-9a9d01fd4f3a

bigint commented:
![Uploading Screenshot 2023-12-18 at 8.49.01 PM.png…]()

make sure to update lock file

bigint commented on apps/web/package.json:
```diff
@@ -65,6 +65,7 @@
     "remark-breaks": "^4.0.0",
     "remark-linkify-regex": "^1.2.1",
     "shared-zustand": "^2.0.0",
+    "solid": "link:@heroicons/react/20/solid",
```
ohh no what is this? 😅

RutamBhagat replied:
Do you want me to replace them with plain svg tags? They are the icons that are used in the select and selected option
![image](https://github.com/heyxyz/hey/assets/72187009/da97b95d-a70b-465f-a4c2-13446cf17dd5)
![image](https://github.com/heyxyz/hey/assets/72187009/2969224e-59d4-448a-a054-20c3fc3beed6)

bigint replied:
we already have heroicon installed, you can use it!

RutamBhagat replied:
done

bigint commented:
<img width="870" alt="Screenshot 2023-12-19 at 12 05 54 AM" src="https://github.com/heyxyz/hey/assets/69431456/7734fa37-7906-4af4-aea3-9045c127752e">

also the input ui and select ui should look same!

bigint commented:
while hover the mouse should use cursor-pointer

bigint commented:
<img width="859" alt="Screenshot 2023-12-19 at 12 06 43 AM" src="https://github.com/heyxyz/hey/assets/69431456/9372f80f-5f72-43a7-a189-797b4cc024fa">

use shadow-sm and radius xl

bigint commented:
<img width="813" alt="Screenshot 2023-12-19 at 12 28 40 AM" src="https://github.com/heyxyz/hey/assets/69431456/72c8c37d-2a01-4837-b958-b482110fecd4">

are we sure both have same shadow?

RutamBhagat commented:
Updated height of select component and changed shadow and corner radius of options menu 

https://github.com/heyxyz/hey/assets/72187009/bc5b9339-9dc0-4a0b-a837-7daf4d1f0fb1

bigint commented:
https://github.com/heyxyz/hey/assets/69431456/7e0d1575-87a7-4418-8a20-198141ecfde1

i dont see which one is selected

bigint commented:
<img width="844" alt="Screenshot 2023-12-19 at 12 29 52 AM" src="https://github.com/heyxyz/hey/assets/69431456/8a42621a-17d4-46b8-98b0-c903c36d8c29">

seeing dupe label

RutamBhagat commented:
> <img alt="Screenshot 2023-12-19 at 12 28 40 AM" width="813" src="https://private-user-images.githubusercontent.com/69431456/291382165-72c8c37d-2a01-4837-b958-b482110fecd4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDI5MjYyMzMsIm5iZiI6MTcwMjkyNTkzMywicGF0aCI6Ii82OTQzMTQ1Ni8yOTEzODIxNjUtNzJjOGMzN2QtMmEwMS00ODM3LWI5NTgtYjQ4MjExMGZlY2Q0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjE4VDE4NTg1M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQwY2JjYTI1NjE0MDRjYmM4N2UzNmFiNmIwODBhNWJmZTU1Nzk0ZDE0ODUzNjViNDNlMTcwMWVmYTkyNjQzNDMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.vRfjGDJxu8t62piANocojR4Q1Mz9Gz1QXqbYEoVLRrU">
> 
> are we sure both have same shadow?

should I remove shadow from the select button?

bigint commented:
<img width="912" alt="Screenshot 2023-12-19 at 12 30 06 AM" src="https://github.com/heyxyz/hey/assets/69431456/a786e7bf-6429-476b-90a5-5d7b31fd7e0b">

fonts should be normal, not sm/xs

bigint commented on packages/ui/src/Select.tsx:
```diff
@@ -1,43 +1,143 @@
+  selected?: boolean;
+  title?: ReactNode;
+  value?: string;
+};
```
```suggestion
};

```

bigint commented:
> should I remove shadow from the select button?

yes

bigint commented:
it should be same as input!

RutamBhagat commented:
> <img alt="Screenshot 2023-12-19 at 12 30 06 AM" width="912" src="https://private-user-images.githubusercontent.com/69431456/291382476-a786e7bf-6429-476b-90a5-5d7b31fd7e0b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDI5MjYzMjMsIm5iZiI6MTcwMjkyNjAyMywicGF0aCI6Ii82OTQzMTQ1Ni8yOTEzODI0NzYtYTc4NmU3YmYtNjQyOS00NzZiLTkwYTUtNWQ3YjMxZmQ3ZTBiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjE4VDE5MDAyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFkNjE5NzU5M2E3NWI5ODRkYzEwMjE0NTJmMzA5ZTZmMzk3ZDM4ZDI1NTJiZDUzYTQ5MDMxMDJjNDQ1M2M2ZmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.vfl35ROqHCgyTn6E38g1yQqR1avUt4mVMmIxmNqwktY">
> 
> fonts should be normal, not sm/xs

![image](https://github.com/heyxyz/hey/assets/72187009/73d74b43-a235-46be-aaed-39257ef91189)
![image](https://github.com/heyxyz/hey/assets/72187009/0fb07775-feaa-4700-a8ac-6725448be3e0)
There is no font-sm or font-xs there are only font-normal, font-medium, and font-semibold to highlight selected items

RutamBhagat commented:
> <img alt="Screenshot 2023-12-19 at 12 29 52 AM" width="844" src="https://private-user-images.githubusercontent.com/69431456/291382408-8a42621a-17d4-46b8-98b0-c903c36d8c29.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDI5MjYzMDMsIm5iZiI6MTcwMjkyNjAwMywicGF0aCI6Ii82OTQzMTQ1Ni8yOTEzODI0MDgtOGE0MjYyMWEtMTdkNC00NmI4LTk4YjAtYzkwM2MzNmQ4YzI5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjE4VDE5MDAwM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWE5NTZmYzM2N2M0NTQ4ZGE2MTYzMGZkYTZhMDI2NGE1ZWI0ZDc2NGY2MzRlYmFhMzVjN2Q0NzFmZWQxOTM4MTgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.N5ocKSvJcL3f4W3TLvHH5k97EIc_1-42pEkBkDKiyJ0">
> 
> seeing dupe label

fixed this, there was a placeholder

RutamBhagat commented:
> Screen.Recording.2023-12-19.at.12.29.07.AM.mov
> 
> i dont see which one is selected

Not sure why this is happening, working on it

RutamBhagat commented:
https://github.com/heyxyz/hey/assets/72187009/45780f88-139c-4150-9b6a-e69e0c6e5d63

@bigint in the Select component i implemented the selected element is shown as active element (It only requires one click) but in the project where the Select components are implemented take double consecutive clicks to change state as shown in the screen recording I am not sure why this is happening can you please suggest me some solution?

bigint commented:
> @bigint in the Select component i implemented the selected element is shown as active element (It only requires one click) but in the project where the Select components are implemented take double consecutive clicks to change state as shown in the screen recording I am not sure why this is happening can you please suggest me some solution?

why we need to use state here? instead select has own props which has selected or not right?

bigint commented:
How is tailwind css example looks like?

RutamBhagat commented:
https://github.com/heyxyz/hey/assets/72187009/84d7a3f0-14b5-4234-beb2-e54f97a5ecb5

@bigint I think I found the issue, this only happens when the options array is placed inside the RFC, when we place the options array outside the RFC the select tag works as expected

RutamBhagat commented:
> How is tailwind css example looks like?

Even in their tailwindcss React example they have defined the people array outside the RFC
![image](https://github.com/heyxyz/hey/assets/72187009/b5ddbb5e-a26a-4172-9ff4-ada573579f5b)

RutamBhagat commented:
https://github.com/heyxyz/hey/assets/72187009/8d034a9a-bf57-4469-9d06-438a8797e136

The Reason.tsx component is working as expected after moving options out of RFC, but in order to do that I had to set every first option as selected: true and rest as false, Should i do this for rest of the components that use Select components?

RutamBhagat commented:
> Screen.Recording.2023-12-19.at.12.29.07.AM.mov
> 
> i dont see which one is selected

another alternative to solve this is to move the options array into useState 

https://github.com/heyxyz/hey/assets/72187009/9eda8c4a-cf96-4b47-a5d7-388a211e340f

bigint commented on apps/web/src/components/Composer/Actions/CollectSettings/AmountConfig.tsx:
```diff
@@ -17,6 +17,12 @@ const AmountConfig: FC<AmountConfigProps> = ({
   allowedTokens,
   setCollectType
 }) => {
+  const [options] = useState(
+    allowedTokens?.map((token) => ({
+      label: token.name,
+      value: token.contractAddress
+    }))
+  );
```
its not a good idea to have state everywhere 😕

RutamBhagat replied:
It causes problems with selection otherwise, the only other option is to move options outside of RFC but that causes problems where the option.selected are dependent on other props

bigint commented:
We need to make this more generic and abstract things on the parent component same like other UI lib does, we can have state inside UI comp instead of all comp using select!

RutamBhagat commented:
> We need to make this more generic and abstract things on the parent component same like other UI lib does, we can have state inside UI comp instead of all comp using select!

![image](https://github.com/heyxyz/hey/assets/72187009/7fd69b07-20a7-4eb9-940e-9af36a9400dc)
yeah it's working after setting new useState inside the Select.tsx with props options as default parameter

bigint commented:
<img width="236" alt="Screenshot 2023-12-20 at 11 44 10 AM" src="https://github.com/heyxyz/hey/assets/69431456/932f96e8-79bd-47e8-bca6-912e44315af2">

padding is off compared to input

RutamBhagat commented:
![image](https://github.com/heyxyz/hey/assets/72187009/355b536b-dc16-44e7-ade2-003a6ca9cc6f)
this is with py-[10px] should i decrease it?

bigint commented:
<img width="911" alt="Screenshot 2023-12-22 at 2 34 23 PM" src="https://github.com/heyxyz/hey/assets/69431456/38c1f909-fcfa-4738-bff3-98641b5cc750">

there is no empty state as well 🙇🏼

RutamBhagat commented:
![image](https://github.com/heyxyz/hey/assets/72187009/4a9d2be8-d30b-4061-8d62-01b038baa475)

what should I provide as default state when no options are provided or are undefined?

RutamBhagat commented:
> We need to make this more generic and abstract things on the parent component same like other UI lib does, we can have state inside UI comp instead of all comp using select!

I have noticed some state problems

https://github.com/heyxyz/hey/assets/72187009/05ac8d61-c0e1-49dd-b2eb-af4a9372075f

its happening because of we moved state inside UI.
when type select is changed the reason sub select options wont change.
I think my previous approach of moving options in the parent component inside useState is the only possible fix for this issue

bigint commented:
https://headlessui.com/react/combobox

bigint commented:
https://headlessui.com/react/listbox

RutamBhagat commented:
> https://headlessui.com/react/listbox

https://github.com/heyxyz/hey/assets/72187009/486a590c-dc67-4b71-b90c-83714a3283bd


I have implemented the components you provided but the selected components are still not showing as active Its exact same issue as earlier. We have to move the options to useState in the parent component or keep it outside of RFC. Passing options directly is causing the selected item being not shown as active and setting useState inside Select.tsx doesn't works either as mentioned earlier

RutamBhagat commented:
> > https://headlessui.com/react/listbox
> 
> simplescreenrecorder-2023-12-23_13.26.00.mp4
> 
> I have implemented the components you provided but the selected components are still not showing as active Its exact same issue as earlier. We have to move the options to useState in the parent component or keep it outside of RFC. Passing options directly is causing the selected item being not shown as active and setting useState inside Select.tsx doesn't works either as mentioned earlier

@bigint should I proceed with moving options in useState in the parent components?
